test/parallel/test-fs-watch-recursive-linux-promise.js

```diff
@@ -0,0 +1,43 @@
+'use strict';
+
+const common = require('../common');
+
+if (!common.hasCrypto)
+  common.skip('missing crypto');
+
+if (common.isIBMi)
+  common.skip('IBMi does not support `fs.watch()`');
+
+const { randomUUID } = require('crypto');
+const assert = require('assert');
+const path = require('path');
+const fs = require('fs/promises');
+
+const tmpdir = require('../common/tmpdir');
+const testDir = tmpdir.path;
+tmpdir.refresh();
+
+(async () => {
+  {
+    // Add a file to already watching folder
+
+    const testsubdir = await fs.mkdtemp(testDir + path.sep);
+    const file = `${randomUUID()}.txt`;
+    const filePath = path.join(testsubdir, file);
+    const watcher = fs.watch(testsubdir, { recursive: true });
+
+    setTimeout(async () => {
+      await fs.writeFile(filePath, 'world');
+    }, 100);
```

Comment:
You need to make sure the file is actually modified, and that it's not the timer keeping the event loop alive but the watcher.
```suggestion
    setTimeout(common.mustCall(() => {
      fs.writeFile(filePath, 'world').then(common.mustCall());
    }), 100).unref();
```

Reply:
Unfortunately, even with these changes, the test fails.

Reply:
Yeah but at least now it doesn't pass if revert to `*[SymbolAsyncIterator](){return this}`. The failure needs to be investigated for sure, it shows that the async iterator doesn't work.